crates/ruff_linter/src/rules/numpy/rules/numpy_2_0_deprecation.rs

```diff
@@ -701,3 +708,240 @@ pub(crate) fn numpy_2_0_deprecation(checker: &mut Checker, expr: &Expr) {
+/// `from numpy import foo` statement, we check to see if that import statement
+/// took place inside an `except ImportError` or `except ModuleNotFoundError` block.
+/// If so, and if the `numpy` member was imported through its non-deprecated name
+/// in the associated try block, we ignore the violation in the same way.
+///
+/// Examples:
+///
+/// ```py
+/// import numpy as np
+///
+/// try:
+///     np.all([True, True])
+/// except AttributeError:
+///     np.alltrue([True, True])  # Okay
+///
+/// try:
+///     from numpy.exceptions import ComplexWarning
+/// except ImportError:
+///     from numpy import ComplexWarning
+///
+/// x = ComplexWarning()  # Okay
+/// ```
+fn is_guarded_by_try_except(
+    expr: &Expr,
+    replacement: &Replacement,
+    semantic: &SemanticModel,
+) -> bool {
+    match expr {
+        Expr::Attribute(_) => {
+            if !semantic.in_exception_handler() {
+                return false;
+            }
+            let Some(try_node) = semantic
+                .current_statements()
+                .find_map(|stmt| stmt.as_try_stmt())
+            else {
+                return false;
+            };
+            let suspended_exceptions = Exceptions::from_try_stmt(try_node, semantic);
+            if !suspended_exceptions.contains(Exceptions::ATTRIBUTE_ERROR) {
+                return false;
+            }
+            try_block_contains_undeprecated_attribute(try_node, &replacement.details, semantic)
+        }
+        Expr::Name(ast::ExprName { id, .. }) => {
+            let Some(binding_id) = semantic.lookup_symbol(id.as_str()) else {
+                return false;
+            };
+            let binding = semantic.binding(binding_id);
+            if !binding.is_external() {
+                return false;
+            }
+            if !binding.in_exception_handler() {
+                return false;
+            }
+            let Some(try_node) = binding
+                .source
+                .and_then(|import_id| enclosing_try_node(import_id, semantic))
+            else {
+                return false;
+            };
+            let suspended_exceptions = Exceptions::from_try_stmt(try_node, semantic);
+            if !suspended_exceptions
+                .intersects(Exceptions::IMPORT_ERROR | Exceptions::MODULE_NOT_FOUND_ERROR)
+            {
+                return false;
+            }
+            try_block_contains_undeprecated_import(try_node, &replacement.details)
+        }
+        _ => false,
+    }
+}
+
+/// Given the `NodeId` of a node that is known to be inside an exception handler,
+/// find the [`ast::StmtTry`] node in which the node resides.
+fn enclosing_try_node<'a>(
+    node_id: NodeId,
+    semantic: &'a SemanticModel,
+) -> Option<&'a ast::StmtTry> {
+    let mut node_id = node_id;
+    loop {
+        node_id = semantic.parent_statement_id(node_id)?;
+        if let ast::Stmt::Try(try_node) = semantic.statement(node_id) {
+            return Some(try_node);
+        }
```

Comment:
Could you instead add a method `statements` to the semantic model that's identical to the `pub fn expressions(&self, node_id: NodeId) -> impl Iterator<Item = &'a Expr> + '_` variant? Or is this doing something else?

Reply:
Ahh, nice idea